### PR TITLE
Per domain quota

### PR DESCRIPTION
**Pull request by mbaechler**
review only last commit

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerDomainMaxQuotaDao.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.quota;
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerDomainMaxQuotaDao.java]: Missing tests for that Dao ?

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaDAO.java [hunk 2]
```diff
@@ -28,6 +28,8 @@
 
 import org.apache.james.mailbox.jpa.quota.model.MaxDefaultMessageCount;
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaDAO.java]: Missing tests for that DAO ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/GenericMaxQuotaManagerTest.java [hunk 3]
```diff
@@ -52,6 +55,14 @@ public void getMaxStorageShouldReturnEmptyWhenNoDefaultValue() throws Exception
         assertThat(maxQuotaManager.getMaxStorage(quotaRoot)).isEmpty();
     }
 
+    @Ignore("how can we link domain and quotaRoot ?")
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/GenericMaxQuotaManagerTest.java]: - This has to be done in `QuotaRootResolver`.

We need IMO to integrate that information to "QuotaRoot" (add it as a field in the object). Then `UserQuotaRootResolver` could simply retrieve it from the user object. (IE resolver becomes responsible to fix the domain, if any).

The problem then becomes "how do we retrieve the field when no resolver is used". IMO we add a field to QuotaRoot.quotaRoot for this.

Note: In non virtualHosting mode, there will be no domain. We have to ensure it still works.

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        this.domainQuotaService = domainQuotaService;
+        this.jsonTransformer = jsonTransformer;
+        this.jsonExtractor = new JsonExtractor<>(QuotaDTO.class, modules.stream().map(JsonTransformerModule::asJacksonModule).collect(Collectors.toList()));
+    }
+
+    @Override
+    public void define(Service service) {
+        this.service = service;
+
+        defineGetQuotaCount();
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: These routes should be defined  ONLY if UserRepository enables virtualHosting.

Else it's a non-sens and using these routes will have no effects.

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaService.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+package org.apache.james.webadmin.routes;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaService.java]: Spotted! Missing license

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerDomainMaxQuotaDao.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            .value(CassandraDomainMaxQuota.MESSAGE_COUNT, bindMarker());
+    }
+
+    private Insert setMaxStorageStatement() {
+        return insertInto(CassandraDomainMaxQuota.TABLE_NAME)
+            .value(CassandraDomainMaxQuota.DOMAIN, bindMarker())
+            .value(CassandraDomainMaxQuota.STORAGE, bindMarker());
+    }
+
+    public void setMaxStorage(String domain, QuotaSize maxStorageQuota) {
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerDomainMaxQuotaDao.java]: I think we are not consistent with domains:
in _CassandraDomainList_ we are using domains in lower case (https://github.com/linagora/james-project/blob/a11ce53c06ad97534334562fc583dab7b73d1294/server/data/data-cassandra/src/main/java/org/apache/james/domainlist/cassandra/CassandraDomainList.java#L114)
not in this dao.

Maybe it's time to have a DAO for domains & introduce domainIds

Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerDomainMaxQuotaDao.java]: Could be done in a separate PR imo.

Until then we can enforce lower case in the different MaxQuotaManagers.

Would that be OK for you?

Comment (aduprat) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerDomainMaxQuotaDao.java]: Ok, just create an issue for that point.

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/DefaultQuotaModule.java [hunk 7]
```diff
@@ -38,7 +38,7 @@ protected void configure() {
         bind(DefaultUserQuotaRootResolver.class).in(Scopes.SINGLETON);
         bind(NoMaxQuotaManager.class).in(Scopes.SINGLETON);
 
-        bind(MaxQuotaManager.class).to(NoMaxQuotaManager.class);
+        bind(MaxQuotaManager.class).toInstance(new NoMaxQuotaManager());
```
Comment (mbaechler) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/DefaultQuotaModule.java]: Previous version was better IMO (or `bind(NoMaxQuotaManager.class).in(Scopes.SINGLETON);` becomes useless)

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java [hunk 8]
```diff
@@ -82,16 +88,26 @@ public DomainQuotaRoutes(DomainList domainList, DomainQuotaService domainQuotaSe
     public void define(Service service) {
         this.service = service;
 
-        defineGetQuotaCount();
-        defineDeleteQuotaCount();
-        defineUpdateQuotaCount();
+        if (isVirtualHostingSupported()) {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: it's not the way I expect things to happen.
I see two solutions : 
1. we decide to expose virtual hosting at guice level so we can inject objects based on its value
2. we always define these roots but we make sure lower level (here `data` module) throws the right exception.
The second one is much easier to discover IMO.

File: server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/MaxQuotaConfigurationReader.java [hunk 9]
```diff
@@ -74,10 +77,14 @@ private void configureDefaultValues(Long defaultMaxMessage, Long defaultMaxStora
...
-            maxQuotaManager.setMaxMessage(QuotaRoot.quotaRoot(entry.getKey()), QuotaCount.count(entry.getValue()));
+            maxQuotaManager.setMaxMessage(toQuotaRoot(entry), QuotaCount.count(entry.getValue()));
         }
         for (Map.Entry<String, Long> entry : maxStorage.entrySet()) {
-            maxQuotaManager.setMaxStorage(QuotaRoot.quotaRoot(entry.getKey()), QuotaSize.size(entry.getValue()));
+            maxQuotaManager.setMaxStorage(toQuotaRoot(entry), QuotaSize.size(entry.getValue()));
         }
     }
+
+    private QuotaRoot toQuotaRoot(Map.Entry<String, Long> entry) throws MailboxException {
```
Comment (mbaechler) [inline server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/MaxQuotaConfigurationReader.java]: passing the key should be enough

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/quota/InMemoryCurrentQuotaManagerTest.java [hunk 10]
```diff
@@ -33,7 +35,7 @@
 
 public class InMemoryCurrentQuotaManagerTest {
 
-    public static final QuotaRoot QUOTA_ROOT = QuotaRoot.quotaRoot("benwa");
+    public static final QuotaRoot QUOTA_ROOT = QuotaRoot.quotaRoot("benwa", Optional.empty());
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/quota/InMemoryCurrentQuotaManagerTest.java]: I don't understand why we use empty almost everywhere

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java [hunk 11]
```diff
@@ -21,17 +21,19 @@
 
 import java.util.Map;
 import java.util.Optional;
+import java.util.function.Function;
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: the commit comment should talk about Quota

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java [hunk 12]
```diff
@@ -133,31 +135,45 @@ public void removeDefaultMaxMessage() {
...
     @Override
     public Optional<QuotaSize> getMaxStorage(QuotaRoot quotaRoot) {
-        return perUserQuota.getMaxStorage(quotaRoot)
-            .map(Optional::of)
-            .orElseGet(Throwing.supplier(this::getDefaultMaxStorage).sneakyThrow());
+        Supplier<Optional<QuotaSize>> domainQuotaSupplier = Throwing.supplier(() -> quotaRoot.getDomain()
+            .flatMap(this::getDomainMaxStorage)).sneakyThrow();
+        Supplier<Optional<QuotaSize>> defaultDomainSupplier = Throwing.supplier(this::getDefaultMaxStorage).sneakyThrow();
+
+        return OptionalUtils.orSuppliers(
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: we are reinventing flatMap.
What about : 
```
Stream.of(() -> perUserQuota.getMaxStorage(quotaRoot), domainQuotaSupplier, defaultDomainSupplier)
  .flatmap(x -> OptionalUtils.toStream(x.get()))
  .findFirst();
```

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java [hunk 13]
```diff
@@ -65,32 +66,32 @@ public void setMaxMessage(QuotaRoot quotaRoot, QuotaCount maxMessageCount) {
 
     @Override
     public void setDomainMaxMessage(String domain, QuotaCount count) {
-        perDomainQuota.setMaxMessage(domain, count);
+        perDomainQuota.setMaxMessage(domain.toLowerCase(Locale.US), count);
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: I'm against this changeset : given the amount of duplication, it's time to introduce a domain object

File: server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/MaxQuotaConfigurationReader.java (hunk 9 above)
Comment (mbaechler) [reply server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/MaxQuotaConfigurationReader.java]: ok

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java (hunk 8 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: ok

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java (hunk 11 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: ok

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java (hunk 12 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: ok

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java (hunk 13 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: I will remove it from this PR and open an new PR about that

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/quota/InMemoryCurrentQuotaManagerTest.java (hunk 10 above)
Comment (mbaechler) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/quota/InMemoryCurrentQuotaManagerTest.java]: Ok, so the answer is : mock everywhere !

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    @PUT
+    @ApiOperation(value = "Updating count and size at the same time")
+    @ApiImplicitParams({
+            @ApiImplicitParam(required = true, dataType = "org.apache.james.webadmin.dto.QuotaDTO", paramType = "body")
+    })
+    @ApiResponses(value = {
+            @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "OK. The value has been updated."),
+            @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "The body is not a positive integer or not unlimited value (-1)."),
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: missing 404

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+        }));
+    }
+
+    @GET
+    @ApiOperation(
+        value = "Reading count and size at the same time",
+        notes = "If there is no limitation for count and/or size, the returned value will be -1"
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = HttpStatus.OK_200, message = "OK", response = QuotaDTO.class),
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: missing 404

Comment (rouazana) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: and everywhere in this file

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+import com.google.common.collect.ImmutableSet;
+import com.jayway.restassured.RestAssured;
+import com.jayway.restassured.http.ContentType;
+import com.jayway.restassured.path.json.JsonPath;
+
+public class DomainQuotaRoutesTest {
+
+    private static final String QUOTA_DOMAINS = "/quota/domains";
+    private static final String PERDU_COM = "perdu.com";
+    private static final String TROUVÉ_COM = "trouvé.com";
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: oh, É is allowed in a variable name in Java, really?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java [hunk 17]
```diff
@@ -247,6 +261,13 @@ public void defineGetQuotaCount() {
     }
 
     private String checkDomainExist(Request request) {
+        if (!isVirtualHostingSupported()) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.METHOD_NOT_ALLOWED_405)
+                .type(ErrorType.WRONG_STATE)
+                .message("Domain Quota configuration not supported")
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: maybe you can explain: "... when virtual hosting is desactivated". Or Even "Please use global quota configuration instead".

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesNoVirtualHostingTest.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    private static final String COUNT = "count";
+    private static final String SIZE = "size";
+    private WebAdminServer webAdminServer;
+
+    @Parameters
+    public static Collection<Object[]> data() {
+        return ImmutableList.of(
+            new String[] {QUOTA_DOMAINS + "/" + FOUND_COM + "/" + COUNT},
+            new String[] {QUOTA_DOMAINS + "/" + FOUND_COM + "/" + SIZE}
+            );
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesNoVirtualHostingTest.java]: you can put this end parenthesis at the end of the previous line

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java [hunk 19]
```diff
@@ -133,31 +135,49 @@ public void removeDefaultMaxMessage() {
 
     @Override
     public Optional<QuotaSize> getMaxStorage(QuotaRoot quotaRoot) {
-        return perUserQuota.getMaxStorage(quotaRoot)
-            .map(Optional::of)
-            .orElseGet(Throwing.supplier(this::getDefaultMaxStorage).sneakyThrow());
+        Supplier<Optional<QuotaSize>> domainQuotaSupplier = Throwing.supplier(() -> quotaRoot.getDomain()
+            .flatMap(this::getDomainMaxStorage)).sneakyThrow();
+        Supplier<Optional<QuotaSize>> defaultDomainSupplier = Throwing.supplier(this::getDefaultMaxStorage).sneakyThrow();
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: defaultDomain? You mean global?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java (hunk 16 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: like emojis

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/DefaultQuotaModule.java (hunk 7 above)
Comment (chibenwa) [reply server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/DefaultQuotaModule.java]: Well but Spring rejected inject on constructors without parameters.... What should we do? Patch Spring?

**Comment by chibenwa** [comment]
```
[71ec7147a283cacf7086a8906e0a5a77baaf362f] Failed tests: 
[71ec7147a283cacf7086a8906e0a5a77baaf362f]   GetQuotaProcessorTest.processorShouldWorkOnExceptionThrown:141 
[71ec7147a283cacf7086a8906e0a5a77baaf362f] Expecting all elements of:
[71ec7147a283cacf7086a8906e0a5a77baaf362f]   <[ImmutableStatusResponse{responseCode=null, serverResponseType=NO, tag=A004, textKey=You need the {0} right to perform command {1} on mailbox {2}., command=Name}]>
[71ec7147a283cacf7086a8906e0a5a77baaf362f] to match given predicate but this element did not:
[71ec7147a283cacf7086a8906e0a5a77baaf362f]   <ImmutableStatusResponse{responseCode=null, serverResponseType=NO, tag=A004, textKey=You need the {0} right to perform command {1} on mailbox {2}., command=Name}>
[71ec7147a283cacf7086a8906e0a5a77baaf362f]   GetQuotaProcessorTest.processorShouldWorkOnValidRights:109 
[71ec7147a283cacf7086a8906e0a5a77baaf362f] responder.respond(<Capturing argument>);
[71ec7147a283cacf7086a8906e0a5a77baaf362f] Wanted 3 times:
[71ec7147a283cacf7086a8906e0a5a77baaf362f] -> at org.apache.james.imap.processor.GetQuotaProcessorTest.processorShouldWorkOnValidRights(GetQuotaProcessorTest.java:109)
[71ec7147a283cacf7086a8906e0a5a77baaf362f] But was 1 time:
[71ec7147a283cacf7086a8906e0a5a77baaf362f] -> at org.apache.james.imap.processor.AbstractMailboxProcessor.no(AbstractMailboxProcessor.java:339)
```

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java (hunk 17 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: Or both!

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java (hunk 19 above)
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: I guess the same vocabulary remarks applies in other max quota manager (memory, jpa)....

Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: As well as everywhere in the code base to make it consistant....

**Comment by chibenwa** [comment]
Ah and we can (**should**) now add integration tests for IsOverQuota and domains <3

**Comment by chibenwa** [comment]
We need to integrate `domain` as part of `QuotaDetailsDTO`

**Comment by chibenwa** [comment]
```
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.messageIdAttachmentIdsShouldMatchBeanContract Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.messageIdAttachmentIdsShouldThrowOnNullAttachmentIds Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.messageIdAttachmentIdsShouldThrowOnNullMessageId Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.retrieveAllMessageIdAttachmentIdsShouldFilterMessagesWithoutAttachment Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.retrieveAllMessageIdAttachmentIdsShouldReturnAllWhenStoredWithAttachment Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.retrieveAllMessageIdAttachmentIdsShouldReturnEmptyWhenNone Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.retrieveAllMessageIdAttachmentIdsShouldReturnEmtpyWhenStoredWithoutAttachment Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.retrieveAllMessageIdAttachmentIdsShouldReturnOneWhenStored Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.retrieveAllMessageIdAttachmentIdsShouldReturnOneWhenStoredWithTwoAttachments Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.saveShouldSaveNullValueForTextualLineCountAsZero Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.saveShouldSaveTextualLineCount Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.saveShouldStoreMessageWithBodyContent Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.saveShouldStoreMessageWithFullContent Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
[a1f9dd693145998d924f6977cf44b75a1641ba2a]   CassandraMessageDAOTest.saveShouldStoreMessageWithHeaderContent Multiple Failures (2 failures)
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	Can't use this cluster instance because it was previously closed
[a1f9dd693145998d924f6977cf44b75a1641ba2a] 	<no message> in java.lang.NullPointerException
```

test this please

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java (hunk 16 above)
Comment (rouazana) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/DomainQuotaRoutesTest.java]: and do we really want to introduce this?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java [hunk 20]
```diff
@@ -116,6 +116,7 @@ public boolean isVirtualHostingSupported() {
     @ApiResponses(value = {
             @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "OK. The value has been updated."),
             @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "The body is not a positive integer or not unlimited value (-1)."),
+            @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "The requested rdomain can not be found."),
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/DomainQuotaRoutes.java]: s/rdomain/domain/ everywhere in this commit

File: server/container/cli-integration/src/test/java/org/apache/james/cli/QuotaCommandsIntegrationTest.java [hunk 21]
```diff
@@ -61,39 +61,39 @@ public void tearDown() {
     }
```
Comment (rouazana) [inline server/container/cli-integration/src/test/java/org/apache/james/cli/QuotaCommandsIntegrationTest.java]: please rename this commit 14f04dfdbcf491c19e5d2cacd92787963c34708d (should be a fixup)

**Comment by chibenwa** [comment]
```
[24ccf1d3045baeef450099face2bdab72b4214e0] domainPutEndpointsShouldReturnNotAllowed[0]  Time elapsed: 0.001 sec  <<< FAILURE!
[24ccf1d3045baeef450099face2bdab72b4214e0] org.opentest4j.MultipleFailuresError: 
[24ccf1d3045baeef450099face2bdab72b4214e0] Multiple Failures (2 failures)
[24ccf1d3045baeef450099face2bdab72b4214e0] 	java.net.UnknownHostException
[24ccf1d3045baeef450099face2bdab72b4214e0] 	<no message> in java.lang.NullPointerException
[24ccf1d3045baeef450099face2bdab72b4214e0] 
```

Maybe linked to the fact reverse  DNS resolution of domains was activated...

File: server/container/cli-integration/src/test/java/org/apache/james/cli/QuotaCommandsIntegrationTest.java (hunk 21 above)
Comment (rouazana) [reply server/container/cli-integration/src/test/java/org/apache/james/cli/QuotaCommandsIntegrationTest.java]: will be done at squash time

**Comment by chibenwa** [comment]
merged

**Pull request closed without merging**